core/src/main/java/org/apache/carbondata/core/statusmanager/StageInputCollector.java

```diff
@@ -0,0 +1,143 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.carbondata.core.statusmanager;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+
+import org.apache.carbondata.common.logging.LogServiceFactory;
+import org.apache.carbondata.core.datastore.filesystem.CarbonFile;
+import org.apache.carbondata.core.datastore.impl.FileFactory;
+import org.apache.carbondata.core.metadata.schema.table.CarbonTable;
+
+import static org.apache.carbondata.core.util.path.CarbonTablePath.SUCCESS_FILE_SUBFIX;
+
+import com.google.gson.Gson;
+import org.apache.commons.io.IOUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.log4j.Logger;
+
+/**
+ * Utilities to create input split from stage files
+ */
+public class StageInputCollector {
+
+  private static Logger LOGGER =
+      LogServiceFactory.getLogService(StageInputCollector.class.getCanonicalName());
+
+  /**
+   * Collect all stage files and create splits from them.
+   * These splits can be included in the queried.
+   */
+  public static List<InputSplit> createInputSplits(CarbonTable table, Configuration hadoopConf)
+      throws ExecutionException, InterruptedException {
+    List<CarbonFile> stageInputFiles = new LinkedList<>();
+    List<CarbonFile> successFiles = new LinkedList<>();
+    collectStageFiles(table, hadoopConf, stageInputFiles, successFiles);
+    if (stageInputFiles.size() > 0) {
+      int numThreads = Math.min(Math.max(stageInputFiles.size(), 1), 10);
+      ExecutorService executorService = Executors.newFixedThreadPool(numThreads);
+      return createInputSplits(executorService, stageInputFiles);
+    } else {
+      return new ArrayList<>(0);
+    }
+  }
+
+  /**
+   * Collect all stage files and matched success files.
+   * A stage file without success file will not be collected
+   */
+  public static void collectStageFiles(CarbonTable table, Configuration hadoopConf,
+      List<CarbonFile> stageInputList, List<CarbonFile> successFileList) {
+    Objects.requireNonNull(table);
+    Objects.requireNonNull(hadoopConf);
+    Objects.requireNonNull(stageInputList);
+    Objects.requireNonNull(successFileList);
+    CarbonFile dir = FileFactory.getCarbonFile(table.getStagePath(), hadoopConf);
+    if (dir.exists()) {
+      // list the stage folder and collect all stage files who has corresponding success file,
+      // which means the file is committed
+      CarbonFile[] allFiles = dir.listFiles();
```

Comment:
i think, better to avoid listfiles to avoid any future problems. it will create many file objects once, we can use nio APIs. like below
`Files.newDirectoryStream(Paths.get(table.getStagePath()), "*.success").forEach(path -> {
        map.put(key, value);
      }
      );`

Reply:
But this way can not access S3, right?

Reply:
can we implement this in our carbonfile API over hadoop conf and try? is it possible?